File: android/app/src/main/java/com/driftless/mapmatch/HmmMapMatcher.kt
package com.driftless.mapmatch

import com.driftless.fusion.MapMatchResult
import com.driftless.fusion.Vec3
import com.driftless.math.minus
import com.driftless.math.norm
import kotlin.math.abs

data class HmmParams(
    val candidateRadiusM: Double = 50.0,
    val maxCandidatesPerObservation: Int = 8,
    val emissionSigmaM: Double = 10.0,
    val transitionBetaM: Double = 5.0,
    val maxRouteDistanceM: Double = 2000.0,
)

/**
 * Hidden Markov Map Matching (Newson & Krumm 2009) against an offline road graph.
 *
 * Implements online causal forward decoding for real-time filter updates [step]
 * and offline full Viterbi with backtracking [match].
 */
class HmmMapMatcher(
    private val graph: RoadGraph,
    private val params: HmmParams = HmmParams(),
) {

    private data class Hypothesis(
        val projection: SegmentProjection,
        var logProbability: Double,
    )

    private var hypotheses = mutableListOf<Hypothesis>()
    private var lastPosition: Vec3 = Vec3()
    private var hasLastPosition = false

    fun reset() {
        hypotheses.clear()
        hasLastPosition = false
    }

    fun emissionLogProbability(distanceM: Double): Double {
        val scaled = distanceM / params.emissionSigmaM
        return -0.5 * scaled * scaled
    }

    fun transitionLogProbability(
        from: SegmentProjection,
        to: SegmentProjection,
        straightLineM: Double,
    ): Double {
        val routeM = graph.routeDistance(from, to, params.maxRouteDistanceM)
        if (routeM < 0.0) return Double.NEGATIVE_INFINITY
        return -abs(straightLineM - routeM) / params.transitionBetaM
    }

    private fun toResult(projection: SegmentProjection): MapMatchResult {
        if (projection.segmentId < 0 || projection.segmentId >= graph.segments.size) {
            return MapMatchResult()
        }
        val seg = graph.segments[projection.segmentId]
        return MapMatchResult(
            matched = true,
            segmentId = projection.segmentId,
            wayId = seg.wayId,
            snappedPosition = projection.point,
            segmentDirection = seg.direction(),
            distanceToRoadM = projection.distanceM,
        )
    }

    /**
     * Causal forward decoding, one observation at a time.
     */
    fun step(position: Vec3): MapMatchResult {
        val candidates = graph.candidatesNear(
            position = position,
            radiusM = params.candidateRadiusM,
            maxCandidates = params.maxCandidatesPerObservation,
        )

        if (candidates.isEmpty()) {
            reset()
            return MapMatchResult()
        }

        val next = mutableListOf<Hypothesis>()

        if (hypotheses.isEmpty() || !hasLastPosition) {
            for (cand in candidates) {
                next.add(Hypothesis(cand, emissionLogProbability(cand.distanceM)))
            }
        } else {
            val straightLine = (position - lastPosition).norm()
            for (cand in candidates) {
                var best = Double.NEGATIVE_INFINITY
                for (prev in hypotheses) {
                    if (prev.logProbability == Double.NEGATIVE_INFINITY) continue
                    val trans = transitionLogProbability(prev.projection, cand, straightLine)
                    if (trans == Double.NEGATIVE_INFINITY) continue
                    val score = prev.logProbability + trans
                    if (score > best) {
                        best = score
                    }
                }
                if (best != Double.NEGATIVE_INFINITY) {
                    next.add(Hypothesis(cand, best + emissionLogProbability(cand.distanceM)))
                }
            }

            if (next.isEmpty()) {
                // Every transition was impossible; restart from emission alone
                for (cand in candidates) {
                    next.add(Hypothesis(cand, emissionLogProbability(cand.distanceM)))
                }
            }
        }

        // Renormalize by subtracting max score
        val bestScore = next.maxOf { it.logProbability }
        if (bestScore.isFinite()) {
            for (hyp in next) {
                hyp.logProbability -= bestScore
            }
        }

        hypotheses = next
        lastPosition = position
        hasLastPosition = true

        val bestHypothesis = hypotheses.maxByOrNull { it.logProbability }
        return if (bestHypothesis != null) toResult(bestHypothesis.projection) else MapMatchResult()
    }

    /**
     * Full offline Viterbi with backtracking over a whole trajectory.
     */
    fun match(positions: List<Vec3>): List<MapMatchResult> {
        val results = MutableList(positions.size) { MapMatchResult() }
        if (positions.isEmpty()) return results

        val candidates = Array(positions.size) { t ->
            graph.candidatesNear(positions[t], params.candidateRadiusM, params.maxCandidatesPerObservation)
        }
        val scores = Array(positions.size) { t ->
            DoubleArray(candidates[t].size) { Double.NEGATIVE_INFINITY }
        }
        val backpointers = Array(positions.size) { t ->
            IntArray(candidates[t].size) { -1 }
        }

        // Forward pass
        var first = positions.size
        for (t in positions.indices) {
            if (candidates[t].isEmpty()) continue
            if (first == positions.size) {
                for (i in candidates[t].indices) {
                    scores[t][i] = emissionLogProbability(candidates[t][i].distanceM)
                }
                first = t
                continue
            }

            var previous = t - 1
            while (previous > first && candidates[previous].isEmpty()) previous--

            val straightLine = (positions[t] - positions[previous]).norm()
            for (i in candidates[t].indices) {
                var best = Double.NEGATIVE_INFINITY
                var bestIdx = -1
                for (j in candidates[previous].indices) {
                    if (scores[previous][j] == Double.NEGATIVE_INFINITY) continue
                    val trans = transitionLogProbability(candidates[previous][j], candidates[t][i], straightLine)
                    if (trans == Double.NEGATIVE_INFINITY) continue
                    val total = scores[previous][j] + trans
                    if (total > best) {
                        best = total
                        bestIdx = j
                    }
                }
                if (best == Double.NEGATIVE_INFINITY) {
                    scores[t][i] = emissionLogProbability(candidates[t][i].distanceM)
                    backpointers[t][i] = -1
                } else {
                    scores[t][i] = best + emissionLogProbability(candidates[t][i].distanceM)
                    backpointers[t][i] = bestIdx
                }
            }

            val bestScore = scores[t].maxOrNull() ?: Double.NEGATIVE_INFINITY
            if (bestScore.isFinite()) {
                for (i in scores[t].indices) {
                    scores[t][i] -= bestScore
                }
            }
        }

        if (first == positions.size) return results

        // Backward pass
        var last = positions.size - 1
        while (last > first && candidates[last].isEmpty()) last--

        var bestLastIdx = 0
        var bestLastScore = Double.NEGATIVE_INFINITY
        for (i in scores[last].indices) {
            if (scores[last][i] > bestLastScore) {
                bestLastScore = scores[last][i]
                bestLastIdx = i
            }
        }

        var index = bestLastIdx
        for (t in last downTo first) {
            if (candidates[t].isEmpty() || index < 0) continue
            results[t] = toResult(candidates[t][index])
            val nextIndex = backpointers[t][index]
            index = if (nextIndex >= 0) {
                nextIndex
            } else {
                var prevT = t - 1
                while (prevT >= first && candidates[prevT].isEmpty()) prevT--
                if (prevT >= first && candidates[prevT].isNotEmpty()) {
                    var bestP = 0
                    var bestPScore = Double.NEGATIVE_INFINITY
                    for (j in scores[prevT].indices) {
                        if (scores[prevT][j] > bestPScore) {
                            bestPScore = scores[prevT][j]
                            bestP = j
                        }
                    }
                    bestP
                } else {
                    -1
                }
            }
        }

        return results
    }
}

File: edge-engine/src/hmm_map_matcher.cpp
#include "driftless/hmm_map_matcher.h"

#include <algorithm>
#include <cmath>
#include <limits>

namespace driftless {
namespace {
// Log-probabilities are only ever compared, never exponentiated, so the
// normalizing constants of both distributions cancel and are omitted.
constexpr double kImpossible = -std::numeric_limits<double>::infinity();
}  // namespace

HmmMapMatcher::HmmMapMatcher(const RoadGraph& graph, const MapMatchParams& params)
    : graph_(&graph), params_(params) {}

double HmmMapMatcher::emissionLogProbability(double distance_m) const {
    const double scaled = distance_m / params_.emission_sigma_m;
    return -0.5 * scaled * scaled;
}

double HmmMapMatcher::transitionLogProbability(const SegmentProjection& from,
                                               const SegmentProjection& to,
                                               double straight_line_m) const {
    const double route_m =
        graph_->routeDistance(from, to, params_.max_route_distance_m);
    // No route within budget: not a low-probability transition but an
    // impossible one. Keeping it as -inf rather than a large penalty stops
    // Viterbi from ever preferring a physically disconnected path.
    if (route_m < 0.0) return kImpossible;
    return -std::abs(straight_line_m - route_m) / params_.transition_beta_m;
}

MapMatchResult HmmMapMatcher::toResult(const SegmentProjection& projection) const {
    MapMatchResult result;
    if (projection.segment_id < 0) return result;
    const RoadSegment& segment = graph_->segments()[projection.segment_id];
    result.matched = true;
    result.segment_id = projection.segment_id;
    result.way_id = segment.way_id;
    result.snapped_position = projection.point;
    result.segment_direction = segment.direction();
    result.distance_to_road_m = projection.distance_m;
    return result;
}

void HmmMapMatcher::reset() {
    hypotheses_.clear();
    has_last_position_ = false;
}

MapMatchResult HmmMapMatcher::step(const Vec3& position) {
    const auto candidates = graph_->candidatesNear(
        position, params_.candidate_radius_m, params_.max_candidates_per_observation);
    if (candidates.empty()) {
        // Off-network: drop the hypothesis set rather than carrying stale
        // beliefs across a gap. Re-acquisition starts clean.
        reset();
        return MapMatchResult{};
    }

    std::vector<Hypothesis> next;
    next.reserve(candidates.size());

    if (hypotheses_.empty() || !has_last_position_) {
        for (const auto& candidate : candidates) {
            next.push_back({candidate, emissionLogProbability(candidate.distance_m)});
        }
    } else {
        const double straight_line = (position - last_position_).norm();
        for (const auto& candidate : candidates) {
            double best = kImpossible;
            for (const auto& previous : hypotheses_) {
                if (previous.log_probability == kImpossible) continue;
                const double transition = transitionLogProbability(
                    previous.projection, candidate, straight_line);
                if (transition == kImpossible) continue;
                best = std::max(best, previous.log_probability + transition);
            }
            if (best == kImpossible) continue;
            next.push_back({candidate, best + emissionLogProbability(candidate.distance_m)});
        }
        if (next.empty()) {
            // Every transition was impossible -- typically a jump across a
            // river or a long GNSS gap. Restart from the emission term alone
            // instead of reporting nothing.
            for (const auto& candidate : candidates) {
                next.push_back({candidate, emissionLogProbability(candidate.distance_m)});
            }
        }
    }

    // Renormalize so scores cannot drift toward -inf over a long trajectory.
    // Only differences matter, so subtracting the maximum changes nothing.
    const double best_score =
        std::max_element(next.begin(), next.end(),
                         [](const Hypothesis& a, const Hypothesis& b) {
                             return a.log_probability < b.log_probability;
                         })
            ->log_probability;
    for (auto& hypothesis : next) hypothesis.log_probability -= best_score;

    hypotheses_ = std::move(next);
    last_position_ = position;
    has_last_position_ = true;

    const auto best = std::max_element(
        hypotheses_.begin(), hypotheses_.end(),
        [](const Hypothesis& a, const Hypothesis& b) {
            return a.log_probability < b.log_probability;
        });
    return toResult(best->projection);
}

std::vector<MapMatchResult> HmmMapMatcher::match(
    const std::vector<Vec3>& positions) const {
    std::vector<MapMatchResult> results(positions.size());
    if (positions.empty()) return results;

    // Full Viterbi: per-observation candidate lists, scores, and backpointers.
    std::vector<std::vector<SegmentProjection>> candidates(positions.size());
    std::vector<std::vector<double>> scores(positions.size());
    std::vector<std::vector<int>> backpointers(positions.size());

    for (std::size_t t = 0; t < positions.size(); ++t) {
        candidates[t] = graph_->candidatesNear(positions[t], params_.candidate_radius_m,
                                               params_.max_candidates_per_observation);
        scores[t].assign(candidates[t].size(), kImpossible);
        backpointers[t].assign(candidates[t].size(), -1);
    }

    // Forward pass.
    std::size_t first = positions.size();
    for (std::size_t t = 0; t < positions.size(); ++t) {
        if (candidates[t].empty()) continue;
        if (first == positions.size()) {
            for (std::size_t i = 0; i < candidates[t].size(); ++i) {
                scores[t][i] = emissionLogProbability(candidates[t][i].distance_m);
            }
            first = t;
            continue;
        }
        // Previous observation that actually had candidates.
        std::size_t previous = t;
        while (previous > first && candidates[previous - 1].empty()) --previous;
        --previous;

        const double straight_line = (positions[t] - positions[previous]).norm();
        for (std::size_t i = 0; i < candidates[t].size(); ++i) {
            double best = kImpossible;
            int best_index = -1;
            for (std::size_t j = 0; j < candidates[previous].size(); ++j) {
                if (scores[previous][j] == kImpossible) continue;
                const double transition = transitionLogProbability(
                    candidates[previous][j], candidates[t][i], straight_line);
                if (transition == kImpossible) continue;
                const double total = scores[previous][j] + transition;
                if (total > best) {
                    best = total;
                    best_index = static_cast<int>(j);
                }
            }
            if (best == kImpossible) {
                // Unreachable from every predecessor: restart the chain here
                // rather than dropping the observation entirely.
                scores[t][i] = emissionLogProbability(candidates[t][i].distance_m);
                backpointers[t][i] = -1;
            } else {
                scores[t][i] = best + emissionLogProbability(candidates[t][i].distance_m);
                backpointers[t][i] = best_index;
            }
        }

        const double best_score = *std::max_element(scores[t].begin(), scores[t].end());
        if (std::isfinite(best_score)) {
            for (auto& score : scores[t]) {
                if (score != kImpossible) score -= best_score;
            }
        }
    }
    if (first == positions.size()) return results;

    // Backward pass from the last observation that had candidates.
    std::size_t last = positions.size() - 1;
    while (last > first && candidates[last].empty()) --last;

    int index = static_cast<int>(std::max_element(scores[last].begin(),
                                                  scores[last].end()) -
                                 scores[last].begin());
    for (std::size_t t = last + 1; t-- > first;) {
        if (candidates[t].empty() || index < 0) continue;
        results[t] = toResult(candidates[t][static_cast<std::size_t>(index)]);
        index = backpointers[t][static_cast<std::size_t>(index)];
        if (index < 0 && t > first) {
            // Chain restarted here; pick up the best candidate of the previous
            // observation on its own merits.
            std::size_t previous = t - 1;
            while (previous > first && candidates[previous].empty()) --previous;
            if (!candidates[previous].empty()) {
                index = static_cast<int>(std::max_element(scores[previous].begin(),
                                                          scores[previous].end()) -
                                         scores[previous].begin());
            }
        }
    }
    return results;
}

}  // namespace driftless
